frontend/src/app/visualization/ngram/ngram.component.ts

```diff
@@ -196,51 +200,46 @@ export class NgramComponent extends ParamDirective implements OnChanges {
     }
 
     loadGraph() {
-        this.isLoading.emit(true);
-
         this.lastParameters = _.clone(this.currentParameters);
         const cachedResult = this.getCachedResult(this.currentParameters);
 
         if (cachedResult) {
             this.onDataLoaded(cachedResult);
         } else {
-            this.visualizationService
-                .getNgramTasks(
-                    this.queryModel,
-                    this.corpus,
-                    this.visualizedField.name,
-                    this.currentParameters
-                )
-                .then((response) => {
-                    this.tasksToCancel = response.task_ids;
-                    return this.apiService.pollTasks<NgramResults>(
-                        response.task_ids
-                    );
-                })
-                .then(([result]) => {
-                    this.tasksToCancel = undefined;
-                    this.cacheResult(result, this.currentParameters);
-                    this.onDataLoaded(result);
-                })
-                .catch(this.onFailure.bind(this));
+            this.visualizationService.getNgramTasks(
+                this.queryModel, this.corpus, this.visualizedField.name,
+                this.currentParameters).then(
+                    response => {
+                        this.tasksToCancel = response.task_ids;
+                        // tasksToCancel contains ids of the parent task and its subtasks
+                        // we are only interested in the outcome of the parent task (first in array)
+                        const poller$ = this.apiService.pollTasks([this.tasksToCancel[0]], this.stopPolling$);
+                        poller$.subscribe({
+                            error: (error) => this.onFailure(error),
+                            next: (result: SuccessfulTask<NgramResults[]>) => this.onDataLoaded((result).results[0]),
+                            complete: () => this.apiService.abortTasks({ task_ids: this.tasksToCancel })
+                    });
+            });
         }
     }
 
-    onFailure(error: { message: string }) {
-        console.log(error);
+    onFailure(error: {message: string}) {
+        this.stopPolling$.next();
+        console.error(error);
         this.currentResults = undefined;
         this.ngramError.emit(error.message);
         this.isLoading.emit(false);
     }
 
     onDataLoaded(result: NgramResults) {
+        this.stopPolling$.next();
```

Comment:
This is a bit confusing because I would think this method is triggered because the poller has finished.

Reply:
No, the result is in, but the poller doesn't stop polling without this.

Reply:
Ah right, I see you removed that logic. Why?

Reply:
Which logic do you refer to? You mean the `take(1)` on `api.service.pollTasks`? With this I reasoned that it would be cleaner to just have the one `takeUntil` operator in the chain.

Reply:
Then again, the `take(1)` operator actually means we don't have to call `stopPolling$.next()` for success/ failure in two components. I think I'll reintroduce it.